core/test/multi_index_test.cpp

```diff
@@ -100,22 +143,122 @@ TEST_F(MultiIndexTest, multiple_data_indices) {
 
 TEST_F(MultiIndexTest, advance_multiple_data_indices) {
   MultiIndex index(yx, x, y);
-  index.advance(1);
+  index.set_index(1);
   check(index, {1, 0, 1, 0, 1}, {0, 1, 1, 2, 2});
-  index.advance(1);
+  index.set_index(2);
   check(index, {0, 1, 0, 1}, {1, 1, 2, 2});
 }
 
 TEST_F(MultiIndexTest, advance_slice_middle) {
   MultiIndex index(xz, xyz);
-  index.advance(2);
+  index.set_index(2);
   check(index, {2, 3, 12, 13, 14, 15});
-  index.advance(3);
+  index.set_index(5);
   check(index, {13, 14, 15});
 }
 
 TEST_F(MultiIndexTest, advance_slice_and_broadcast) {
   MultiIndex index(xz, xy);
-  index.advance(2);
+  index.set_index(2);
   check(index, {0, 0, 3, 3, 3, 3});
 }
+
+TEST_F(MultiIndexTest, 1d_array_of_1d_buckets) {
+  const Dim dim = Dim::Row;
+  Dimensions buf{{dim}, {7}}; // 1d cut into two sections
+  // natural order no gaps
+  check_with_buckets(buf, dim, {{0, 3}, {3, 7}}, x, x, {0, 1, 2, 3, 4, 5, 6});
+  // gap between
+  check_with_buckets(buf, dim, {{0, 3}, {4, 7}}, x, x, {0, 1, 2, 4, 5, 6});
+  // gap at start
+  check_with_buckets(buf, dim, {{1, 3}, {3, 7}}, x, x, {1, 2, 3, 4, 5, 6});
+  // out of order
+  check_with_buckets(buf, dim, {{4, 7}, {0, 4}}, x, x, {4, 5, 6, 0, 1, 2, 3});
+}
+
+TEST_F(MultiIndexTest, 1d_array_of_2d_buckets) {
+  Dimensions buf{{Dim("a"), Dim("b")}, {2, 3}}; // 2d cut into two sections
+  // cut along inner
+  check_with_buckets(buf, Dim("b"), {{0, 1}, {1, 3}}, x, x, {0, 3, 1, 2, 4, 5});
+  check_with_buckets(buf, Dim("b"), {{0, 1}, {2, 3}}, x, x, {0, 3, 2, 5});
+  check_with_buckets(buf, Dim("b"), {{1, 2}, {2, 3}}, x, x, {1, 4, 2, 5});
+  check_with_buckets(buf, Dim("b"), {{1, 3}, {0, 1}}, x, x, {1, 2, 4, 5, 0, 3});
+  // cut along outer
+  check_with_buckets(buf, Dim("a"), {{0, 1}, {1, 2}}, x, x, {0, 1, 2, 3, 4, 5});
+  check_with_buckets(buf, Dim("a"), {{1, 2}, {1, 2}}, x, x, {3, 4, 5, 3, 4, 5});
+  check_with_buckets(buf, Dim("a"), {{1, 2}, {0, 1}}, x, x, {3, 4, 5, 0, 1, 2});
+}
```

Comment:
Do you have a test that checks if you overrun the size, e.g.
```C++
check_with_buckets(buf, Dim("a"), {{0, 1}, {1, 3}}, x, x, {0, 1, 2, 3, 4, 5}); // should throw error
```

Reply:
No, because it does not. Similar to `std::vector::operator[]` we do not have bounds check here. This would harm performance. However, provided that there are no bugs in `transform` (and we have tests for that) this should never happen, because `transform` checks the `dims` of the inputs before beginning with the iteration.